Fix kubelet certificate reload when connecting by IP address

kwohlfahrt opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

Note the issue is tagged as a feature, but certificate reloading for the kubelet was previously added in #124574, so I think this is better described as a bug.

#### What this PR does / why we need it:

Currently, the certificate reloading functions from the `ReloadKubeletServerCertificateFile` are not called. A clearer description is in [this issue comment](https://github.com/kubernetes/kubernetes/issues/133001#issuecomment-3209843823), but the summary is that `http.Server` does not call `GetCertificate` if certificate paths were passed to `ListenAndServeTLS` and the client connects by IP (and therefore does not use SNI).

#### Which issue(s) this PR is related to:

Fixes #133001 

#### Special notes for your reviewer:

I'm struggling to find a good way to test for the presence/absence of this bug, because `ListenAndServeKubeletServer` calls `os.Exit(1)` if the HTTPS server exits. There is also no easy way to shutdown the HTTPS server, or to pick a random port to serve on (for parallel testing).

I'm happy to refactor this code to allow testing, but would like to get a maintainer/reviewer's opinion on how to best do this - should it be an e2e test, or I change the function to `CreateKubeletServer`, and return the server for testing?

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubelet: fixed reloading of kubelet server certificate files when they are changed on disk, and kubelet is dialed by IP address instead of DNS/hostname
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

N/A

SergeyKanzhelev commented:
Are there any tests that should have failed, but didn't? Or new tests that you can write?

SergeyKanzhelev commented:
/cc @zhangweikop, the author of the referred PR

SergeyKanzhelev commented:
> @kwohlfahrt Thank you for debugging and fixing the issue.
> 
> I think using e2e test to covert this makes sense. I feel refactoring might not help much, since the logic still ultimately depends on the behavior of Go's tls.Server.

We need to have some test here to catch a regression. @zhangweikop @kwohlfahrt if you will have troubles identifying the proper way to test it, please comment here

kwohlfahrt committed:
Fix kubelet certificate reload when connecting by IP

Currently, we set TLSConfig.Config.GetCertificate, but then also pass
certificate and key paths to http.Server.ListenAndServeTLS.

ListenAndServeTLS uses these paths to populate the TLS config Certificate
property. Then, when accepting connections, a non-nil Certificate is preferred
over GetCertificate if the ServerName is not set in ClientHelloInfo. Finally,
the Go TLS client doesn't set ServerName when connecting by IP. As a result,
when connecting to the kubelet by IP (e.g. to fetch pod logs), stale
certificates are served.

This patch passes empty certFile and keyFile arguments, to force the TLS
server to use the GetCertificate function.

This is done by clearing key/cert file config when setting GetCertificate as
suggested in PR review. This way, all downstream users of kubeDeps.TLSConfig
will do the right thing automatically.

stlaz commented on pkg/kubelet/server/server.go:
```diff
@@ -194,10 +194,17 @@ func ListenAndServeKubeletServer(
-		// cert/keys are specified and GetCertificate in the TLSConfig
-		// should be called instead.
-		if err := s.ListenAndServeTLS(tlsOptions.CertFile, tlsOptions.KeyFile); err != nil {
+		certFile := tlsOptions.CertFile
```
I believe a better place to clear the Cert/KeyFile would be at the spot where the `GetCertificate()` function is being set: https://github.com/kubernetes/kubernetes/blob/e1773f1afd16d7e2d9e0cd9b2c853734abb76ea5/pkg/kubelet/kubelet.go#L913-L921

Here it may not be clear where `tlsOptions` come from, and if `tlsOptions.CertFile` and `tlsOptions.KeyFile` would also be returned in a `GetCertificate()` call.
On the other hand, if we clear these fields in the above mentioned code, all consumers should continue to use all of `Cert/KeyFile` and `config.GetCertificate()` for their serving configuration.

kwohlfahrt replied:
Yeah, I didn't do this initially because I thought it might be confusing to have `tlsOptions.CertFile` and `tlsOptions.KeyFile` set in `InitializeTLS`, and then unset again later in `NewMainKubelet`. But I agree it might be less error prone to unset them immediately.

I've added the change, (+ a rebase onto `master`)

kwohlfahrt commented:
/retest

I'm pretty sure the failure is not from this PR, but from kubernetes/test-infra#35882

SergeyKanzhelev commented:
/skip

stlaz commented on pkg/kubelet/kubelet.go:
```diff
@@ -919,6 +919,13 @@ func NewMainKubelet(ctx context.Context,
 				return cert, nil
 			}
+
+			// GetCertificate is only preferred over the certificate files if
```
```suggestion
			// GetCertificate is only preferred over the certificate files by Golang TLS if
```

stlaz reviewed (commented):
minor comments, lgtm

stlaz reviewed (approved):
/lgtm
/approve

liggitt commented:
/lgtm
/approve
/milestone v1.36
